homeassistant/helpers/state.py

```diff
@@ -132,6 +99,103 @@ def get_changed_since(states: Iterable[State],
             if state.last_updated >= utc_point_in_time]
 
 
+@bind_hass
+def async_map_state_to_calls(hass: HomeAssistantType,
+                             domain: str,
+                             state: State,
+                             state_to_service: Dict[str, str],
+                             services: Dict[str,
+                                            voluptuous.schema_builder.Schema]
+                             ) -> List[ServiceCall]:
+        """
+        Maps a state object into service calls to reproduce the given state.
+
+        Function will make two passes. First it will create a call to reproduce
+        the state.state, passing along any allowed attributes to the state
+        setting service. As a second pass it will map remaining attributes
+        to services
+        """
+
+        calls = []
+
+        # expanded attribute list containing entity id
+        expanded_attributes = dict(state.attributes)
+        expanded_attributes[ATTR_ENTITY_ID] = state.entity_id
+
+        def add_if_complete(service: str) -> None:
+            """Add a service call with the attributes needed."""
+            schema = voluptuous.schema_builder.Schema(
+                services[service].schema,
+                required=services[service].required,
+                extra=voluptuous.schema_builder.REMOVE_EXTRA)
+
+            try:
+                attributes = schema(expanded_attributes)
+            except voluptuous.error.Invalid as e:
+                _LOGGER.debug("Incomplete attribute "
+                              "list to call {}:{}".format(service, e))
+                return
+
+            # consume attributes taken by service
+            # this avoid multiple service calls
+            # for same attribute if schema would
+            # allow it
+            for key in attributes:
+                expanded_attributes.pop(key, None)
```

Comment:
Is this a problem we've encountered or is it just theoretical? As far as I know, there are no shared attributes between services for a domain (for services that are not directly affecting a state, eg on/off).

I think we could just assume that we'll continue following that rule, and make it a rule when defining service attributes. Ie attribute names should be unique per domain for services that only modify state attributes.

Reply:
light.turn_on light.toggle share transition.

Reply:
climate.set_temperature climate.set_operation_mode share operation mode.

Reply:
We will never call turn on and toggle at the same time, right?

It's weird that `set_temperature` takes operation mode. We should probably look at removing that for the future. But for now we can handle that in the climate component.

Reply:
`set_temperature` also requires other attributes, so only operation_mode won't trigger both services.

Reply:
nope, but if you do set a temperature and give it operation mode, you get two calls.

Reply:
> We will never call turn on and toggle at the same time, right?
For state restore it would if we don't consume the transition when calling turn on. The schema for toggle would still match.

Reply:
But the schema shouldn't both validate turn on and toggle. This is why we should make new schemas in each component.

Reply:
It sounds good to combine state and state attributes into a single validation schema, since we don't want to call more than one service that changes state but we can allow many services that changes different attributes.

About your example, I'm not sure that `transition` without state should be allowed. The transition attribute is connected to a service that should change the state. So we should require the state in the scene.